### PR TITLE
Rewriting the mutator code with numpy arrays

### PR DESCRIPTION
**Pull request by teytaud**
## Types of changes

- [ ] Docs change / refactoring / dependency upgrade
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist

- [x] The documentation is up-to-date with the changes I made.
- [x] I have read the [**CONTRIBUTING**](https://facebookresearch.github.io/nevergrad/contributing.html) document and completed the CLA (see [**CLA**](https://facebookresearch.github.io/nevergrad/contributing.html#contributor-license-agreement-cla)).
- [x] All tests passed, and additional code has been covered with new tests.

### CODE REVIEW

File: nevergrad/optimization/mutations.py [hunk 1]
```diff
@@ -66,26 +66,26 @@ def portfolio_discrete_mutation(self, parent: tp.ArrayLike, intensity: tp.Option
             intensity = 1 if dimension == 1 else int(self.random_state.randint(1, dimension))
         if dimension == 1:  # corner case.
             return self.random_state.normal(0., 1., size=1)  # type: ignore
-        boolean_vector = [True for _ in parent]
+        boolean_vector = np.array(np.ones(dimension), dtype=bool)
```
Comment (jrapin) [inline nevergrad/optimization/mutations.py:69]:
```suggestion
        boolean_vector = np.ones(dimension, dtype=bool)
```

File: nevergrad/optimization/mutations.py [hunk 2]
```diff
@@ -66,26 +66,26 @@ def portfolio_discrete_mutation(self, parent: tp.ArrayLike, intensity: tp.Option
...
-            boolean_vector = [self.random_state.rand() > (float(intensity) / dimension) for _ in parent]
+            boolean_vector = self.random_state.rand(dimension) > float(intensity) / dimension
         return [s if b else self.significantly_mutate(s, arity) for (b, s) in zip(boolean_vector, parent)]
 
     def coordinatewise_mutation(self, parent: tp.ArrayLike, velocity: tp.ArrayLike, boolean_vector: tp.ArrayLike, arity: int) -> tp.ArrayLike:
         """This is the anisotropic counterpart of the classical 1+1 mutations in discrete domains
         with tunable intensity: it is useful for anisotropic adaptivity."""
         dimension = len(parent)
-        boolean_vector = [False for _ in parent]
+        boolean_vector = np.array(np.zeros(dimension), dtype=bool)
```
Comment (jrapin) [inline nevergrad/optimization/mutations.py:78]:
```suggestion
        boolean_vector = np.zeros(dimension, dtype=bool)
```

File: nevergrad/optimization/mutations.py [hunk 3]
```diff
@@ -66,26 +66,26 @@ def portfolio_discrete_mutation(self, parent: tp.ArrayLike, intensity: tp.Option
...
+            boolean_vector = self.random.rand(dimension) < (1. / dimension)
         discrete_data = discretization.threshold_discretization(parent, arity=arity)
         return discretization.inverse_threshold_discretization([s if not b else s + np.random.choice([-1. ,1.]) * v for (b, s, v) 
                                                                 in zip(boolean_vector, discrete_data, velocity)])
     
     def discrete_mutation(self, parent: tp.ArrayLike, arity: int = 2) -> tp.ArrayLike:
         """This is the most classical discrete 1+1 mutation of the evolution literature."""
         dimension = len(parent)
-        boolean_vector = [True for _ in parent]
+        boolean_vector = np.array(np.ones(dimension), dtype=bool)
```
Comment (jrapin) [inline nevergrad/optimization/mutations.py:88]:
```suggestion
        boolean_vector = np.ones(dimension, dtype=bool)
```

File: nevergrad/optimization/mutations.py [hunk 4]
```diff
@@ -66,26 +66,26 @@ def portfolio_discrete_mutation(self, parent: tp.ArrayLike, intensity: tp.Option
...
 
     def coordinatewise_mutation(self, parent: tp.ArrayLike, velocity: tp.ArrayLike, boolean_vector: tp.ArrayLike, arity: int) -> tp.ArrayLike:
         """This is the anisotropic counterpart of the classical 1+1 mutations in discrete domains
         with tunable intensity: it is useful for anisotropic adaptivity."""
         dimension = len(parent)
-        boolean_vector = [False for _ in parent]
+        boolean_vector = np.array(np.zeros(dimension), dtype=bool)
         while not any(boolean_vector):
-            boolean_vector = [self.random_state.rand() < (1. / dimension) for _ in parent]
+            boolean_vector = self.random.rand(dimension) < (1. / dimension)
```
Comment (jrapin) [inline nevergrad/optimization/mutations.py:80]:
```suggestion
            boolean_vector = self.random_state.rand(dimension) < (1. / dimension)
```

**Comment by jrapin** [comment]
Definitely a good habit to take, thanks ;)

File: nevergrad/optimization/mutations.py (hunk 1 above)
Comment (teytaud) [reply nevergrad/optimization/mutations.py:69]: How I could come up with something THAT stupid ...

File: nevergrad/optimization/mutations.py (hunk 4 above)
Comment (teytaud) [reply nevergrad/optimization/mutations.py:80]: Twice... :-/

File: nevergrad/optimization/mutations.py (hunk 3 above)
Comment (teytaud) [reply nevergrad/optimization/mutations.py:88]: 3 times...

File: nevergrad/optimization/mutations.py (hunk 2 above)
Comment (teytaud) [reply nevergrad/optimization/mutations.py:78]: 4 times...

**Comment by jrapin** [comment]
please merge master inside this coordinatewise-adaptivity otherwise we cannot figure out the changes

**Comment by jrapin** [comment]
this branch and the one it is based on are using different master, so it's impossible to compare. Maybe just merge and I'll review `coordinatewise-adaptivity` branch

**Comment by teytaud** [comment]
I close it for now and I'll see that after merging 

> this branch and the one it is based on are using different master, so it's impossible to compare. Maybe just merge and I'll review `coordinatewise-adaptivity` branch

I'm scared, so I propose that we merge coordinatewise-adaptivity first and I'll check what is going on in the present PR and/or just repeat from scratch, which should be easy.

**Comment by teytaud** [comment]
Closed until merging coordinatewise-adaptivity.

**Comment by jrapin** [comment]
> I close it for now and I'll see that after merging
> 
> > this branch and the one it is based on are using different master, so it's impossible to compare. Maybe just merge and I'll review `coordinatewise-adaptivity` branch
> 
> I'm scared, so I propose that we merge coordinatewise-adaptivity first and I'll check what is going on in the present PR and/or just repeat from scratch, which should be easy.

I'm reopening to make sure we don't lose this. I change the base branch of this to master. You can merge coordinatewise-adaptivity, and then merge possible conflicts

**Comment by jrapin** [comment]
notice that the diff is way smaller now that the base branch is master, the issue is just that coordinate-stuff is not up-to-date

**Review by jrapin — APPROVED** [review]

**Pull request merged**